File: src/numerical_pdes/mesh.py
import numpy as np
from abc import ABC, abstractmethod
from numpy.typing import NDArray


class Mesh(ABC):
    @property
    @abstractmethod
    def ndim(self) -> int:
        ...

    @property
    @abstractmethod
    def n_cells(self) -> int:
        ...


class Mesh1D(Mesh):
    def __init__(self, x_left: float, x_right: float, n_cells: int):
        self._n_cells = n_cells
        self.x_left = x_left
        self.x_right = x_right
        self.dx: float = (x_right - x_left) / n_cells
        self.cell_centers: NDArray = np.linspace(
            x_left + self.dx / 2,
            x_right - self.dx / 2,
            n_cells,
        )

    @property
    def ndim(self) -> int:
        return 1

    @property
    def n_cells(self) -> int:
        return self._n_cells

    @property
    def period(self) -> float:
        return self.x_right - self.x_left

File: src/numerical_pdes/csrc/mesh.cpp
#include "mesh.h"


double Mesh1d::GetPeriod() {
    return x_right_ - x_left_;
}


torch::Tensor Mesh1d::GetCellCentersWithGhostCells(int n_ghost_cells) const {
    auto n_cells = cell_centers_.size(0);
    auto far_left = cell_centers_.index({0}) - n_ghost_cells*dx_;
    auto far_right = cell_centers_.index({n_cells - 1}) + n_ghost_cells*dx_;

    torch::Tensor x = torch::linspace(
                far_left,
                far_right,
                n_cells + 2 * n_ghost_cells,
                torch::kFloat64);
    return x;
}
